### DISCUSSION TITLE
Coupled Microstrip Line

### DISCUSSION THREAD

**Discussion by CosimoMV** [HelgeGehring/femwell#135] [Q&A]
I would like to calculate the electric and magnetic field for the even and odd eigenmodes of a coupled microstrip.
Before I start studying the documentation in depth, I would like to know if Femwell is capable of doing such an analysis.

**Comment by HelgeGehring** [comment]
Hey @CosimoMV ,

right now we don't have yet an example for coplanar waveguides, but it would be great to add one!

The physics is the same as in the optical domain, so it shouldn't be a problem to do that analysis using Femwell.
I've started in https://github.com/HelgeGehring/femwell/blob/main/femwell/examples/coplanar_waveguide.py (which currently simulates a coax cable).
The main difference is that we have to think more about the boundary conditions.

I'm more in the optical domain, but happy to learn more about electronics.
The main reason that I didn't go on with that example is that I don't have a paper which we could reproduce to make sure we're doing things right.

Would you have a paper (or book) which shows even and odd eigenmodes? Along with numbers such as the effective index which we could match? :)

**Reply by CosimoMV** [reply]
First of all, thank you for your reply!

> Would you have a paper (or book) which shows even and odd eigenmodes? Along with numbers such as the effective index which we could match? :)

Although not very recent, a good source for summarising various information is "Microstrip Lines and Slotlines" third edition published by Artech House, especially section 8.3.2 and 8.3.3. 
The best way to compare the correctness of the modes found would be to use some kind of mode-solver that does not use a quasi-static approximation, which unfortunately I do not have. 
For frequencies around 1GHz, if one wants to limit oneself to calculating impedance and effective electrical permittivity, it is very convenient to use Kikad's "Calculator Tools", which is open source and free.

> The main difference is that we have to think more about the boundary conditions.

I am not familiar with finite element methods, but for FDFD (finite difference frequency domain) mode solvers, the boundary conditions are very problematic for the microstrip.
There are many papers on this, but sections 2.1 and 2.2 of this [paper](https://ars.copernicus.org/articles/18/7/2020/) summarise several issues.

I have personally tried using Tidy3D's free FDFD mode solver, but at frequencies around 1GHz I only get non-physical modes both with PEC boundary and using PML. Only at 10GHz, and using a really large simulation zone, I was able to obtain a physical mode.

**Reply by CosimoMV** [reply]
I copied and pasted various pieces of code from the examples on the Femwell's site, to arrive at something that more or less resembled a microstrip.

This is the code:

```
from collections import OrderedDict
import matplotlib.pyplot as plt
import numpy as np
import scipy.constants
from scipy.constants import epsilon_0, speed_of_light
from scipy.integrate import solve_ivp
from shapely.geometry import Polygon
from skfem import Basis, ElementTriP0, Mesh
from skfem.io import from_meshio
from femwell.maxwell.waveguide import compute_modes
from femwell.mesh import mesh_from_OrderedDict
from femwell.utils import inside_bbox
mils=25.4
#dl=0.05*mils
w_sin=4.55*mils
w_si = w_sin
gap=5.26*mils
#ground_under_h=5*dl
#dx_lat=ground_under_h
#dy_lat=ground_under_h
h_box=79.1
h_ground=26.3+h_box
h_sin=53.4
h_si=h_sin
w_sim = 7.*(h_sin+h_si+gap)
h_clad=(h_box+h_sin)*5.


#wavelength = 30*
#k0 = 2 * np.pi / wavelength

polygons = OrderedDict(
    sin=Polygon(
        [
            (-w_sin - gap / 2, 0),
            (-w_sin - gap / 2, h_sin),
            (-gap / 2, h_sin),
            (-gap / 2, 0),
        ]
    ),
    si=Polygon(
        [
            (w_si + gap / 2, 0),
            (w_si + gap / 2, h_si),
            (gap / 2, h_si),
            (gap / 2, 0),
        ]
    ),
    clad=Polygon(
        [
            (-w_sim / 2, 0),
            (-w_sim / 2, h_clad),
            (w_sim / 2, h_clad),
            (w_sim / 2, 0),
        ]
    ),
    box=Polygon(
        [
            (-w_sim / 2, 0),
            (-w_sim / 2, -h_box),
            (w_sim / 2, -h_box),
            (w_sim / 2, 0),
        ]
    ),
    ground=Polygon(
        [
            (-w_sim / 2, 0),
            (-w_sim / 2, -h_ground),
            (w_sim / 2, -h_ground),
            (w_sim / 2, 0),
        ]
    )
)

resolutions = dict(
    sin={"resolution": 0.2*mils, "distance": 1},
    si={"resolution": 0.2*mils, "distance": 1},
)

mesh = from_meshio(
    mesh_from_OrderedDict(polygons, resolutions, filename="mesh.msh", default_resolution_max=0.2*mils)
)
#mesh.draw().show()
basis0 = Basis(mesh, ElementTriP0(), intorder=4)
f = 1e9
wavelength=speed_of_light/(f)
epsilon = basis0.zeros().astype(complex)
epsilon[basis0.get_dofs(elements="box")] = 3.6+ 1j *0.02
epsilon[basis0.get_dofs(elements="ground")] = 1. + 1j * 5.8e7 * 1e20 / f
epsilon[basis0.get_dofs(elements="clad")] = 1.0
epsilon[basis0.get_dofs(elements="sin")] = 1. + 1j * 5.8e7 * 1e20 / f
epsilon[basis0.get_dofs(elements="si")] = 1. + 1j * 5.8e7 * 1e20 / f
modes = compute_modes(
        basis0,
        epsilon,
        wavelength=wavelength,
        mu_r=1,
        num_modes=3,
        #metallic_boundaries=True,
        n_guess=np.sqrt(3.2),
        order=1
    )
#mesh.draw().show()
for mode in modes:
    mode.show(mode.E.real,colorbar=True)
```

I played with the size of the micro-strip's section, the resolution, order, and metallic_boundaries, but I did not obtain any physical mode.

**Reply by HelgeGehring** [reply]
Yay, great!

What about exacly rebuilding a geometry from the book you mentioned? The one on page 453? Then we could stepwise get closer to solving it :)

Do you know how usually the conductors are modeled? Would we use a material with a high conductivity as in that example? Or would we just cut it out of the simulation and use perfectly electric conductors as a boundary condition?
Or put in a highly conducting strip but surround it with a PEC?

I've also found https://www.tandfonline.com/doi/abs/10.1080/02726340290084012 where they describe their exact geometry (based on that is the code you found and thus the number 5.8e7. We could also work towards solving that one (or both :) )

**Reply by CosimoMV** [reply]
> What about exacly rebuilding a geometry from the book you mentioned? The one on page 453?

On page 453 is shown an asymmetric microstrip line and I think is better to start with a symmetric one, that should be simpler. By the way, there is a way to consider symmetries with Femwell? If so, we could use the mirror symmetry about y-z plane to discriminate odd and even modes.

> Do you know how usually the conductors are modeled? Would we use a material with a high conductivity as in that example? Or would we just cut it out of the simulation and use perfectly electric conductors as a boundary condition?

The copper metal planes and traces of PCBs have roughness (which can also be high depending on the type) and furthermore, the traces are not rectangular, but trapezoidal. All this changes the conductivity (there are also some models), but these are higher order corrections. For now, I would first try to find odd and even modes of an "ideal microstrip" and then add the various corrections to make the model more realistic. So, for now, I think is ok to use a "generic" high conductivity material for metal.

For the boundary condition I don't know. Either we use a large enough area so that the fields are almost zero at the boundary, or else I fear that using PEC will introduce non-physical artefacts. I would like to try with PMC if they could be implemented on Femwell.

> I've also found https://www.tandfonline.com/doi/abs/10.1080/02726340290084012 where they describe their exact geometry (based on that is the code you found and thus the number 5.8e7. We could also work towards solving that one (or both :) )

I used that specific stripline geometry in the code I posted because it is a real microstrip of a PCB whose physical parameters were measured by the manufacturer. Furthermore, I had managed to find a mode at 10 GHz (or at least a mode that seemed plausible and not blatantly an artefact) using Tidy3D's FDFD solver and wanted to see if I could recreate the same result with Femwell. 
Anyway, thanks for the paper! I'll try to implement that geometry now and see if I can get something meaningful.

**Reply by HelgeGehring** [reply]
I tried to implement the example on page 453, but they didn't mention the thickness of the conductor, which makes it hard to reproduce :/

Mirror symmetry is not a problem, depending on which modes you want, you just need a PEC or PMC boundary as a symmetry axis :)

Ok, sounds interesting! Maybe we could add those corrections also to the docs and explain them :smiley: 

@lucasgrjn mentioned https://ieeexplore.ieee.org/document/1129316 which has all the parameters which are needed for reproducing. I've added them to the docs! The coplanar waveguide is already online. https://helgegehring.github.io/femwell/electronics/examples/coplanar_waveguide_vary_gap.html#, the strip waveguide should follow once the docs are rebuilt, probably in ~2h :)
(you can already find it under docs/electronics/examples in the repo)

The values don't match yet exactly the paper but it's pretty close. Could you check if you find something we should change to make it fit better?
(If you want you could also add some text explaining the example :) I don't know yet how to interpret the even/odd mode for a RF waveguides :/ but happy to learn :) )
For those simulations FEM is probably the better suited method as the length-scales vary a lot.

I'd guess the examples are also a good starting point for your simulation :)

**Reply by CosimoMV** [reply]
Wow! Thank you very much! 
I just saw the two new examples and they seem very useful. I got stuck in some meetings and I am only now starting to look at the code and the papers you linked.

> The values don't match yet exactly the paper but it's pretty close. Could you check if you find something we should change to make it fit better?

I will definitely try to figure out what the differences are due to (If I can) and how to introduce the copper roughness.

> I don't know yet how to interpret the even/odd mode for a RF waveguides :/ but happy to learn :) )

Sorry if I didn't explain well regarding the odd and even modes. Odd modes (they are what I am really interested in) are obtained with a PEC symmetry about y-z plane and they are depicted on page 443 of the book. Odd modes are interesting because when using microstrips for digital signaling, the two lines are generally driven differentially (with signals of the same amplitude and opposite polarity), so, hopefully, only the odd modes are excited.

**Reply by HelgeGehring** [reply]
Sounds great!

Are there other examples we should add to the docs? I'm wondering if Fig. 4 in https://ieeexplore.ieee.org/document/1129316 would also be something interesting?

Yay, great! Let me know if we need to adjust something! Would the rest of the figures also be something relevant to add?

Thanks for that explanation! Helps a lot :)

Is there something like an often in literature discussed geometry? (In photonics those explanations often start with a single mode strip waveguide at 1550nm either in silicon or in silicon nitride as those are the most commonly used ones)
I'd be happy if we could add such an example to the docs (with plots of even/odd modes, ....) and explain it well :)

**Reply by CosimoMV** [reply]
> I'm wondering if Fig. 4 in https://ieeexplore.ieee.org/document/1129316 would also be something interesting?

In my opinion the graph would be interesting. The most important thing shown by the graph (although it would be useful to have a few more points) is the presence of the cut-off frequency for the first excited mode. If someone were to use the microstrip for a real application, the signalling frequency should be well below the cutoff frequency. 

> Is there something like an often in literature discussed geometry? 

Unfortunately, to my knowledge, there isn't such a thing.

I looked a bit at how to add the surface roughness of the copper, but it would be necessary to consider a surface impedance for both the microstrip and the copper foil at the bottom, and I have no idea how to do this with Femwell. Some very useful references can be found in these two papers: Roughness Characterization for Interconnect Analysis by Yuriy Shlepnev and Chudy Nwachukwu published in 2011 IEEE International Symposium on Electromagnetic Compatibility and [this paper](https://www.signalintegrityjournal.com/ext/resources/article-images-2018/2018a/4040_ACausalConductorRoughnessModelPDF.pdf) from Signal Integrity Journal.

As for the discrepancies between the results obtained in the paper and those from Femwell, I think they are due to the different methodologies and the approximations made. For example, in the paper the starting point is to consider the field E tangential to the infinitely thin strip to be zero. Then the actual thickness of the strip is considered as a perturbation via formulae 8 and 9. It seems to me that the discrepancies concern strips with a smaller width, for which I do not know how applicable the approximations made in the paper are.

**Reply by CosimoMV** [reply]
I also have a more general question concerning FEM solvers. My ultimate goal would be to find the S-parameters of two coupled microstrips connected by a via (a metallised hole connecting two or more planes of a PCB) so that I can optimise certain characteristics of the circuit. My initial idea was to find the modes of the miscrostrip and use an FDTD simulator (Meep) to simulate the complete structure. Basically following the procedure explained [here](https://meep.readthedocs.io/en/latest/Mode_Decomposition/), but for metal structures for which MPB cannot be used. However, I realised that this is not feasible, because the mesh for metal structures should be too fine to be able to simulate something real. I think it would be more realistic to use a FEM simulator. I think Femwell is not usable in this case because these are not structures homogeneous in one direction, but could you advise me on a feasible way?

**Reply by HelgeGehring** [reply]
> In my opinion the graph would be interesting. The most important thing shown by the graph (although it would be useful to have a few more points) is the presence of the cut-off frequency for the first excited mode. If someone were to use the microstrip for a real application, the signalling frequency should be well below the cutoff frequency.

Okay, then let's add it :) what about the other figures?

> I looked a bit at how to add the surface roughness of the copper, but it would be necessary to consider a surface impedance for both the microstrip and the copper foil at the bottom, and I have no idea how to do this with Femwell. Some very useful references can be found in these two papers: Roughness Characterization for Interconnect Analysis by Yuriy Shlepnev and Chudy Nwachukwu published in 2011 IEEE International Symposium on Electromagnetic Compatibility and [this paper](https://www.signalintegrityjournal.com/ext/resources/article-images-2018/2018a/4040_ACausalConductorRoughnessModelPDF.pdf) from Signal Integrity Journal.

Surface impedance sounds like a line term we'd need to add to the weakforms for the mode solver :thinking: what exacly does it describe? an additional impedance because of two materials being connected?
The roughness sounds like something people include more empirical and not directly in the geometry? I.e. after the mode solve? Or are there ways to put it in the geometry (/epsilons)? Could we just lower the conductivity on the outside of the structures? I.e. exponentially let the value go from the conductivity of the solid to 0 over the RMS depth?
In https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9864759/ they give a skinning depth and a roughtness depth and it seems the more the mode is concentraded at the surface the more it's impacted :thinking: 
@simbilod did an example which considers it similar in the optical domain - he adjusted the epsilon at the outside and that gave us a model for the propagation loss. Might be similar here?
Thanks for the references, I'll have a look!

> As for the discrepancies between the results obtained in the paper and those from Femwell, I think they are due to the different methodologies and the approximations made. For example, in the paper the starting point is to consider the field E tangential to the infinitely thin strip to be zero. Then the actual thickness of the strip is considered as a perturbation via formulae 8 and 9. It seems to me that the discrepancies concern strips with a smaller width, for which I do not know how applicable the approximations made in the paper are.

Ok, sounds perfect, then it's totally expected that we have different results! Especially for the smaller widths those approximations are expected to have problems as they usually rely on thickness << width
I'll add a note to the example!

> I also have a more general question concerning FEM solvers. My ultimate goal would be to find the S-parameters of two coupled microstrips connected by a via (a metallised hole connecting two or more planes of a PCB) so that I can optimise certain characteristics of the circuit. My initial idea was to find the modes of the miscrostrip and use an FDTD simulator (Meep) to simulate the complete structure. Basically following the procedure explained [here](https://meep.readthedocs.io/en/latest/Mode_Decomposition/), but for metal structures for which MPB cannot be used. However, I realised that this is not feasible, because the mesh for metal structures should be too fine to be able to simulate something real. I think it would be more realistic to use a FEM simulator. I think Femwell is not usable in this case because these are not structures homogeneous in one direction, but could you advise me on a feasible way?

Yeah, FEM sounds more reasonable for that!
Up to now we don't have it implementet (yet!) - but we have several necessary parts.
- The mode solver is needed to launch modes and capture them
- I started working on FEFD to do those simulations, but there's some stuff missing, in principal launching/capturing modes and the PMLs are note yet implemented.
We should puzzle it more together :) 
Would you have a paper where they simulated that which we could try to reproduce?

**Reply by CosimoMV** [reply]
> Okay, then let's add it :) what about the other figures?

In my opinion the other interesting figures are: 7,9 and 10.

The roughness of the copper is due to the PCB manufacturing process. Generally, it is something that decreases the performance of the PCB (in fact, there are more and more expensive processes and construction methods to decrease these effects as much as possible), but it also makes the copper microstrip or sheet "stick" to the dielectric. In fact, it also serves to prevent delamination of the PCB when subjected to high stress (such as in avionics or even certain rolling stock tests). Manufacturers are very jealous of their production processes and tend not to divulge much information, but what they certainly do say (and which corresponds to the truth, unlike the declared /epsilon_r values) is the RMS value of roughness. In the two papers I linked to you, they refer precisely to models that can be described by an RMS value. At high frequencies (above GHZ), the skin effect causes the current to be distributed over a few micrometres on the conductor surface, so values that can also be much lower than the RMS value of the roughness (well-developed case).
One of the models described in the papers describes the Hammerstad correction coefficient which takes into account losses (so reletatd to the real part of impedence) due to roughness. The problem is that in order to maintain causality, the inductive contribution of the impedance must also be taken into account, and therefore one must choose appropriately the coefficients that multiply the real and imaginary part of impedance (this issue is also addressed in the first part of the second paper).
But even if one were to consider modifying the impedance value per  unit length found for a smooth transmission line (TEM modes) and correcting it by an appropriate factor, the fact remains that the microstrip supports QUASI-TEM modes, which deviate from TEM modes as the frequency increases... so I don't think this is a viable route in this specific case (like in the paper you linked). In the first paper I linked there are are proposed some possible solutions but they do not seem easy to implement. I hope that what I wrote make sense!
> Would you have a paper where they simulated that which we could try to reproduce?

I don't have a paper describing measurements or simulations concerning two coupled microstrips connected by a via, but I will definitely look for it! I will write to you as soon as I find it.
